File: algorithms/Unique Paths III/Unique Paths III.py
class Solution:
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        ans, empty = 0, 1
        
        def dfs(grid: List[List[int]], row: int, col: int, count: int, visited) -> None:
            if row >= len(grid) or col >= len(grid[0]) or row < 0 or col < 0 or grid[row][col] == -1:
                return
            nonlocal ans
            if grid[row][col] == 2:
                if empty == count:
                    ans += 1
                return
            if (row, col) not in visited:
                visited.add((row, col))
                dfs(grid, row + 1, col, count + 1, visited)
                dfs(grid, row - 1, col, count + 1, visited)
                dfs(grid, row, col + 1, count + 1, visited)
                dfs(grid, row, col - 1, count + 1, visited)
                visited.remove((row, col))
        
        row, col = 0, 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    row, col = i, j
                elif grid[i][j] == 0:
                    empty += 1
        dfs(grid, row, col, 0, set())
        return ans

File: scripts/algorithms/U/Unique Paths III/Unique Paths III.js
// Runtime: 79 ms (Top 83.70%) | Memory: 42 MB (Top 97.04%)
// 980. Unique Paths III
var uniquePathsIII = function(grid) {
    const M = grid.length; // grid height
    const N = grid[0].length; // grid width
    let result = 0; // final result
    let startY = 0, startX = 0; // starting point coordinates
    let finalY = 0, finalX = 0; // endpoint coordinates
    let empty = 0; // count of empty squares
    let visit = Array(M); // visited squares (MxN of booleans)

    // Initialization of required variables
    for (let i = 0; i < M; i++) {
        visit[i] = Array(N).fill(false); // now: "visit[i][j] === false"
        for (let j = 0; j < N; j++) {
            switch (grid[i][j]) {
                case 0 : empty++; break;
                case 1 : startY = i; startX = j; break;
                case 2 : finalY = i; finalX = j; break;
            }
        }
    }

    // Recursively run DFS and get the answer in the "result" variable
    dfs(startY, startX, visit, 0);
    return result;

    // DFS implementation
    function dfs (startY, startX, visit, step) {

        // If it's a wrong square, then exit immediately
        if (startY < 0 || startY >= M || // off grid (height)
            startX < 0 || startX >= N || // off grid (width)
            visit[startY][startX] || // already processed
            grid [startY][startX] === -1 // this is an obstacle
        ) return; // ... exit now

        // Base case: we're at the endpoint, need to stop the recursion.
        // If all of squares are visited, increase the "result":
        // (count of paths from start to the end).
        if (startY === finalY && startX === finalX) {
            if (step - 1 === empty) result++;
            return;
        }

        // Run DFS for neighboring squares.
        // Increase the number of steps (count of the visited squares).
        visit[startY][startX] = true; // mark current square as visited
        dfs(startY-1, startX, visit, step + 1); // top
        dfs(startY, startX+1, visit, step + 1); // right
        dfs(startY+1, startX, visit, step + 1); // bottom
        dfs(startY, startX-1, visit, step + 1); // left
        visit[startY][startX] = false; // restore visited square
    }
};